File: Binary Search Tree Iterator/Binary_Search_Tree_Iterator.cpp
// LeetCode上有人分析了，这种方式其实是满足题目对于时间复杂度和空间复杂度的要求的
// Runtime: 60 ms, faster than 98.85% of C++ online submissions for Binary Search Tree Iterator.
// Memory Usage: 24.4 MB, less than 97.41% of C++ online submissions for Binary Search Tree Iterator.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator 
{
public:
    BSTIterator(TreeNode* root) 
    {
        push(root);
    }
    
    /** @return the next smallest number */
    int next() 
    {
        TreeNode* tempNode = nodeStack.top();
        nodeStack.pop();

        if (tempNode->right)
            push(tempNode->right);
        
        return tempNode->val;
    }
    
    /** @return whether we have a next smallest number */
    bool hasNext() 
    {
        return !nodeStack.empty();
    }
private:
    stack<TreeNode*> nodeStack;

    void push(TreeNode* root)
    {
        while (root)
        {
            this->nodeStack.push(root);
            root = root->left;
        }
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: Binary Search Tree Iterator/Binary_Search_Tree_Iterator.py
# Runtime: 108 ms, faster than 23.36% of Python3 online submissions for Binary Search Tree Iterator.
# Memory Usage: 20.2 MB, less than 100.00% of Python3 online submissions for Binary Search Tree Iterator.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:

    def __init__(self, root: 'TreeNode'):
        self.nodeStack = []
        self.push(root)

    def next(self) -> 'int':
        """
        @return the next smallest number
        """
        tempNode = self.nodeStack.pop()
        if tempNode.rihgt:
            self.push(tempNode.right)
        return tempNode.val
                
    def hasNext(self) -> 'bool':
        """
        @return whether we have a next smallest number
        """
        return not self.nodeStack

    def push(self, root):
        while root:
            self.nodeStack.append(root)
            root = root.left
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()